source/common/network/proxy_protocol.cc

```diff
@@ -61,17 +60,49 @@ void ProxyProtocol::ActiveConnection::onReadWorker() {
     return;
   }
 
-  if (proxy_line.find(PROXY_TCP4) != 0) {
+  // Parse proxy protocol line with format: PROXY TCP4/TCP6 SOURCE_ADDRESS DESTINATION_ADDRESS
+  // SOURCE_PORT DESTINATION_PORT.
+  const auto line_parts = StringUtil::split(proxy_line, " ", true);
+
+  if (line_parts.size() != 6 || line_parts[0] != "PROXY") {
+    throw EnvoyException("failed to read proxy protocol");
+  }
+
+  Address::IpVersion protocol_version;
+  if (line_parts[1] == "TCP4") {
+    protocol_version = Address::IpVersion::v4;
+  } else if (line_parts[1] == "TCP6") {
+    protocol_version = Address::IpVersion::v6;
+  } else {
     throw EnvoyException("failed to read proxy protocol");
   }
 
-  size_t index = proxy_line.find(" ", PROXY_TCP4.size());
-  if (index == std::string::npos) {
+  // Error check the source and destination fields. Currently the source port, destination address,
+  // and destination port fields are ignored. Remote address refers to the source address.
+  Address::InstanceConstSharedPtr remote_address = Utility::parseInternetAddress(line_parts[2]);
+  Address::InstanceConstSharedPtr destination_address =
+      Utility::parseInternetAddress(line_parts[3]);
+  const auto remote_version = remote_address->ip()->version();
+  const auto destination_version = destination_address->ip()->version();
+  if (remote_version != protocol_version || destination_version != protocol_version) {
     throw EnvoyException("failed to read proxy protocol");
   }
 
-  size_t addr_len = index - PROXY_TCP4.size();
-  std::string remote_address = proxy_line.substr(PROXY_TCP4.size(), addr_len);
+  uint32_t remote_port, destination_port;
+  try {
+    remote_port = std::stoul(line_parts[4]);
+    destination_port = std::stoul(line_parts[5]);
+    // Check that TCP port value is in range [0, 65535].
+    if (remote_port > 65535 || destination_port > 65535) {
+      throw std::out_of_range("failed to read proxy protocol");
+    }
+  } catch (const std::invalid_argument& ia) {
```

Comment:
@hennna FWIW, I'm not a huge fan of this type of error checking of the C++ runtime functions by catching std exceptions. It can be really unclear what throws and why, and we might catch something that we don't intend. It's not a huge deal, but in this case, would rather see us just verify things using standard if length checks, and then using StringUtil::atoul which is a lot faster than the standard version. Not a big deal at all but we might consider a follow up. cc @htuch